### PR TITLE
[Perf] [MOSS-TTS] Add breakable Prefill CUDA Graph support

### PR DESCRIPTION
**Pull request by charliechenye**
## Motivation

MOSS TTS builds request specific `input_embeds` during prefill. Writing those embeddings directly to `ForwardBatch.input_embeds` makes request specific state visible during graph admission, so MOSS TTS cannot enter the shared breakable Prefill CUDA Graph path.

This PR adopts the existing Omni prefill sidecar instead. Request specific embeddings stay outside `ForwardBatch.input_embeds` during admission and are late-bound through the shared `SGLModelRunner` forward path after graph eligibility is decided.

This is intentionally a small adopter PR. It does not add a MOSS-specific graph runner and does not change MOSS TTS generation math, sampling, decode CUDA Graphs, or the vocoder. Breakable prefill remains **opt-in**.

## Changes

- Mark MOSS TTS as compatible with breakable Prefill CUDA Graph in the model capability and engine builder.
- Route request specific prefill embeddings through `OmniPrefillInputs` instead of `ForwardBatch.input_embeds`.
- Accept the shared `omni_prefill_rids` forward kwarg required by the Omni prefill transport contract.
- Add focused tests for opt-in policy and MOSS-specific sidecar attachment.

The shared `SGLModelRunner` tests continue to cover sidecar-to-forward-kwargs transport.

### CPU / contract

- MOSS prefill sidecar and transport: **6 passed**
- Model capability contract: **36 passed**
- Full MOSS unit suite: **142 passed, 12 skipped**
- Shared prefill infrastructure: **37 passed**
- Pre-commit and patch-scope checks passed

### Correctness and quality

Real SeedTTS requests exercised both graph replay and eager fallback while keeping `ForwardBatch.input_embeds` unset during graph admission.

A full 1,088-sample SeedTTS EN quality comparison showed no observed aggregate regression:

| Metric | Prefill graph OFF | Prefill graph ON |
|---|---:|---:|
| Corpus WER | **2.0514%** | **2.0514%** |
| Speaker similarity mean | 68.9292 | 69.0850 |
| Evaluated | 1088 / 1088 | 1088 / 1088 |

A current-head H200 follow-up also compared eager, exact-shape BCG, and production-bucket-padded BCG on representative 150, 179, and 183-token prefills. Captured pre-sampling hidden states were bit-exact in all comparisons, and generated WAV SHA-256 matched across all three execution modes. Channel logits were not exposed at the qualification hook, so no direct channel-logit parity claim is made.

### Performance

A matched concurrency-1 profiler isolated the targeted prefill phase:

| Metric | OFF | ON | Change |
|---|---:|---:|---:|
| Prefill mean host-wall | 24.08 ms | 13.97 ms | **-41.99%** |
| Prefill p50 | 22.66 ms | 13.91 ms | **-38.59%** |
| Prefill p95 | 31.00 ms | 15.18 ms | **-51.02%** |
| Ready-state graph memory | | | ~**+270 MiB** |

The three-round counterbalanced H100 end-to-end campaign covered **20,784 measured requests with zero request failures**:

| Concurrency | Median paired QPS change | Median paired mean-latency change |
|---:|---:|---:|
| 1 | +0.34% | -0.35% |
| 8 | +2.76% | -2.70% |
| 16 | **+4.04%** | **-3.86%** |
| 32 | **+4.75%** | **-4.53%** |

c16 and c32 improved throughput in all three measured rounds. c8 was more variable and included one regression round, so this PR does not claim a universal end-to-end improvement at every concurrency.

### Replay / fallback coverage

The current-head H200 coverage pass used the audited historical MOSS serving contract and measured one full ON-only workload at each concurrency:

| Concurrency | Graph request coverage | Graph useful-token coverage |
|---:|---:|---:|
| 1 | 90.00% | 99.90% |
| 8 | 98.53% | 99.19% |
| 16 | 92.74% | 92.12% |
| 32 | 73.62% | 72.85% |

All eager fallbacks were explained by the shared policy: either the realized scheduled-token batch exceeded the 512-token capture envelope or the next capture bucket would require more than 2x padding. There were no unattributed fallback cases.

The full 30-bucket production ladder through 512 captured on H200 in about **3.8-4.0 s**, with SGLang reporting **0.31 GiB** capture memory. A matched OFF/ON startup check measured **+264 MiB** ready-state HBM with BCG enabled and left about **39.4 GiB (27.4%)** free. No startup-time speedup claim is made.

## Scope

This PR only makes MOSS TTS compatible with the shared breakable Prefill CUDA Graph path.

It intentionally does **not**:

- enable breakable prefill by default;
- prescribe one bucket ladder for every deployment;
- change decode CUDA Graph behavior;
- change sampling or generation semantics;
- change vocoder behavior.

The current evidence supports breakable prefill as an **opt-in optimization**. Default-on policy can be evaluated separately.

## Related issue

Related to #1357. This PR covers the MOSS TTS adopter only and does not close the broader tracker.

## Checklist

- [x] Format code according to pre-commit.
- [x] Add focused unit tests.
- [x] Validate real graph replay and eager fallback.
- [x] Validate output quality.
- [x] Provide prefill-stage and multi-concurrency performance evidence.
- [x] Provide current-head replay/fallback coverage and capture/headroom qualification.
- [ ] For reviewers: If you have not made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the `run-ci` label. Once labeled, every subsequent push re-triggers CI as long as the label remains.

Use `/tag-and-rerun-ci moss` to select MOSS-TTS CI.

### CODE REVIEW

**Review by AkazaAkane — CHANGES_REQUESTED** [review]
Code looks good. But we need more validation: please add deterministic exact-shape-vs-padded mismatch attribution; report benchmark replay/fallback coverage + scheduled token distribution; complete capture/startup/headroom record.

File: tests/unit_test/moss_tts/test_pipeline.py [hunk 1]
```diff
@@ -1575,14 +1576,99 @@ def _prepare_multi_modal_inputs(self, rows):
...
+        hidden_size=2,
+        pp_group=SimpleNamespace(is_first_rank=True, is_last_rank=True),
+        model=backbone,
+        _prepare_multi_modal_inputs=lambda rows: rows.to(torch.float32)[:, :2],
+        _select_sample_hidden_states=lambda hidden_states, forward_batch: hidden_states,
+    )
+    moss_runner = MossTTSModelRunner.__new__(MossTTSModelRunner)
+    moss_runner.model = model
+    sched_req = SimpleNamespace(
+        data=SimpleNamespace(
```
Comment (stephenkgli) [inline tests/unit_test/moss_tts/test_pipeline.py]: After merging the current `main`, this test fails before reaching the transport assertions because `MossTTSModelRunner._build_prefill_input_embeds()` reads `data.output_rows`, which this `SimpleNamespace` does not define.

Comment (charliechenye) [reply tests/unit_test/moss_tts/test_pipeline.py]: Fixed, the fixture now includes output_rows=[] to match the current MOSS prefill state

**Comment by charliechenye** [comment]
> Code looks good. But we need more validation: please add deterministic exact-shape-vs-padded mismatch attribution; report benchmark replay/fallback coverage + scheduled token distribution; complete capture/startup/headroom record.

@AkazaAkane  Thanks for the review. I completed the requested current-head H200 follow-up qualification for MOSS-TTS Prefill BCG.

Qualification was run on PR head `bbbe0b11119220152ec9d72e9c55b2e85b82c49f`, with the pinned MOSS-TTS model, codec, and SeedTTS EN dataset. The serving contract was audited against the historical H100 benchmark setup before the GPU runs: `max_running_requests=64`, `mem_fraction_static=0.65`, preprocessing and vocoder behavioral settings left at the model defaults, and the production Prefill BCG ladder captured through 512 tokens.

## 1. Deterministic exact-shape vs padded attribution

I compared the same logical request under three execution modes:

1. eager at the raw shape
2. BCG with an exact capture bucket
3. BCG with the next production capture bucket

Current head representatives were:

| Raw scheduled tokens | Exact BCG | Production padded BCG |
|---:|---:|---:|
| 150 | 150 | 160 |
| 179 | 179 | 192 |
| 183 | 183 | 192 |

For all three representatives:

- eager vs exact BCG captured hidden states were bit exact
- exact BCG vs padded BCG captured hidden states were bit exact
- maximum absolute error was `0`
- generated WAV SHA 256 matched across eager, exact, and padded execution

So I could not reproduce a deterministic numerical divergence from either graph replay or production-bucket padding on H200.

The historical H100 observations around 180/186 tokens are treated only as references to the same 192-token production bucket. The current PR head produced nearby 179/183 token representatives instead.

Channel logits were not exposed at the qualification hook, so I am not making a direct channel-logit parity claim.

## 2. Replay/fallback coverage and scheduled-token distribution

I then ran one full ON only coverage characterization under the audited historical serving/traffic contract, including 20 sequential warmup requests before each measured workload.

All `3,464` measured requests completed successfully with `0` request failures.

| Offered concurrency | Measured requests | Graph batch coverage | Graph request coverage | Graph useful-token coverage | Scheduled tokens p50 / p95 / max |
|---:|---:|---:|---:|---:|---:|
| c1 | 200 | 90.00% | 90.00% | 99.90% | 130.5 / 168.1 / 190 |
| c8 | 1088 | 98.67% | 98.53% | 99.19% | 147 / 325.6 / 586 |
| c16 | 1088 | 96.94% | 92.74% | 92.12% | 275.5 / 448.1 / 924 |
| c32 | 1088 | 86.83% | 73.62% | 72.85% | 293 / 667.8 / 1090 |

Fallback attribution was fully accounted for:

| Concurrency | Padding gap fallback requests | `scheduled_tokens > 512` fallback requests | Other / unattributed |
|---:|---:|---:|---:|
| c1 | 20 | 0 | 0 |
| c8 | 9 | 7 | 0 |
| c16 | 1 | 78 | 0 |
| c32 | 1 | 286 | 0 |

The c1 percentage needs one nuance: the 20 eager requests were one token realized prefills after the sequential warmup reused the first dataset entries. They account for only 20 eager useful tokens versus 20,356 graph useful tokens, which is why request coverage is 90.0% while useful token coverage is 99.90%.

At higher concurrency, the dominant fallback is aggregate scheduler batching above the 512 token capture envelope. For example, c32 reaches a 1,090 token realized prefill batch. This is expected eager fallback from the shared SGLang policy, not a MOSS specific failure.

Most successful graph replays use padding rather than an exact capture shape:

| Concurrency | Exact graph batches | Padded graph batches |
|---:|---:|---:|
| c1 | 16 | 164 |
| c8 | 53 | 765 |
| c16 | 28 | 573 |
| c32 | 19 | 403 |

This also makes the deterministic exact vs padded check representative of the normal BCG execution path rather than a rare edge case.

Padding overhead over graph served useful tokens was about 6.90% at c1, 6.05% at c8, 5.31% at c16, and 4.81% at c32. There were no unexplained eager fallbacks.

## 3. Capture/startup/headroom

I also reran the startup/capture measurement with matched OFF and ON server configurations. The only treatment difference was Prefill BCG:

- OFF: `cuda_graph_backend_prefill=disabled`
- ON: `cuda_graph_backend_prefill=breakable` with the full 30 bucket production ladder through 512

Results on H200:

| Metric | Result |
|---|---:|
| Full production capture ladder | 30 buckets through 512 |
| Capture elapsed | 3.80 s |
| SGLang reported capture memory | 0.31 GiB |
| Capture available memory | 46.03 GiB -> 45.71 GiB |
| Ready state HBM delta, ON vs OFF | +264 MiB |
| HBM free after BCG startup | 39,444 MiB |
| Free HBM fraction | 27.44% |
| Process to health OFF | 38.03 s |
| Process to health ON | 38.01 s |

I do not interpret the approximately `-0.02 s` process to health difference as a startup speedup. The direct capture elapsed/memory and matched ready state HBM delta are the useful overhead records.

## Conclusion

The three requested follow ups are now covered:

- deterministic exact shape vs padded attribution: no mismatch reproduced
- replay/fallback coverage plus scheduled token distribution: characterized at c1/c8/c16/c32 with all fallbacks explained
- capture/startup/headroom: full 512 token ladder captured successfully with a few hundred MiB of resident HBM overhead and substantial remaining H200 headroom

I am keeping the previously reported counterbalanced H100 campaign as the performance evidence. The H200 coverage run above is ON only qualification, so I am not using its QPS or latency numbers as a new OFF/ON performance comparison.

**Review by AkazaAkane — APPROVED** [review]
LGTM
